### ISSUE TITLE
pretraining loss blows up with Triton Flash attention

### ISSUE THREAD

**Issue by yzhang123** [Dao-AILab/flash-attention#337]
when pretraining gpt with triton flash attention loss blows up (from ~2 to 7) halfway into the training and doesn't go down anymore. If i resume from a healthy ckpt without Flash attention the loss is stable.
I could reproduce this error.

I was using Alibi, bf16 O2 precision

**Comment by tridao** [comment]
The Triton implementation is experimental, I did see some race conditions from the Triton compiler on the backward pass (see comments in the source code) that I tried to fix. Maybe not all are fixed.
Maybe things are better with the new Triton compiler backend? I haven't had the bandwidth to play around too much.

**Comment by mayankjobanputra** [comment]
I am facing the same issue, gonna try to switch to the torch implementation. Will report my findings here.

**Comment by yzhang123** [comment]
@mayankjobanputra would highly appreciate it!

**Comment by mayankjobanputra** [comment]
@yzhang123 I was able to avoid the loss explosion by increasing the warm-up steps 3x than default and changing the learning rate scheduler to linear decay (with warm up). Maybe you can try it too.

**Comment by yzhang123** [comment]
thanks!

**Comment by EricLina** [comment]
Same problem.

**Comment by T4t4KAU** [comment]
I would like to know the latest developments on this issue and what caused it. Thank you!
